### PR TITLE
Awesome IPLD endeavour

### PR DESCRIPTION
**Pull request by diasdavid**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -2,7 +2,7 @@
   "name": "ipfs-unixfs-engine",
   "version": "0.11.4",
   "description": "JavaScript implementation of the unixfs Engine used by IPFS",
-  "main": "lib/index.js",
+  "main": "src/index.js",
```
Comment (dignifiedquire) [inline package.json]: TODO: remove before merge

File: src/importer/index.js [hunk 2]
```diff
@@ -3,14 +3,14 @@
 const UnixFS = require('ipfs-unixfs')
 const assert = require('assert')
 const pull = require('pull-stream')
-const pushable = require('pull-pushable')
-const write = require('pull-write')
+const pullPushable = require('pull-pushable')
```
Comment (dignifiedquire) [inline src/importer/index.js]: why this renaming?

Comment (diasdavid) [inline src/importer/index.js]: it helps to know that 'it is a pull stream thing'

**Comment by diasdavid** [comment]
I've migrated the importer to the Async interfaces and had to go through a major overhaul of the `flush-tree.js`, since it was relying on it being very sync to traverse the tree.

unixfs-engine needs an overhaul cleanup, refactor and simplification, it takes a bit to sync in what is going on inside the code, although the high level algorithm is quite simple. flush-tree.js has improved a lot with this PR.

I'm just missing 4 last tests for the importer, namely the ones with directories, traverse is getting called twice for the same dir, increasing the level of nestness, still trying to understand why. Left some logs to help with the debugging.

**Comment by dignifiedquire** [comment]
thank you @diasdavid! lets wait for the full refactoring until we add the hamt though, as that will add additional constraints on the interfaces

**Comment by dignifiedquire** [comment]
@diasdavid fixed and rebase

**Comment by diasdavid** [comment]
@dignifiedquire ❤️

**Pull request merged**